src/coreComponents/constitutive/fluid/PVTFunctions/CO2Solubility.cpp

```diff
@@ -125,40 +126,42 @@ real64 Par( real64 const & T, real64 const & P, real64 const * cc )
   return x;
 }
 
-real64 CO2SolubilityFunction( real64 const & tolerance,
+real64 CO2SolubilityFunction( string const & name,
+                              real64 const & tolerance,
                               real64 const & T,
                               real64 const & P,
                               real64 (* f)( real64 const & x1, real64 const & x2, real64 const & x3 ) )
 {
-  constexpr integer maxIter = 50;
-  constexpr real64 dx = 1e-10;
-
-  real64 Vr_int = 0.05;
-  real64 V_r = 0.75*Rgas*(T_K_f+T)/(P*P_Pa_f)*(1/V_c);
-
-  // iterate until the solution of the CO2 equation of state is found
-  integer count = 0;
-  real64 dre = LvArray::NumericLimits< real64 >::infinity;
-  for(; count < maxIter && fabs( dre ) >= tolerance; ++count )
-  {
-    if( V_r < 0.0 )
-    {
-      V_r = Vr_int;
-      Vr_int += 0.05;
-    }
-    real64 const v0 = (*f)( T, P, V_r );
-    real64 const v1 = (*f)( T, P, V_r+dx );
-    dre = -v0/((v1-v0)/dx);
-    V_r += dre;
-  }
-
-  GEOSX_THROW_IF( count == maxIter,
-                  "CO2Solubility NR convergence fails! " << "dre = " << dre << ", tolerance = " << tolerance,
-                  InputError );
-  return V_r;
+  // compute the initial guess for Newton's method
+  real64 const initialReducedVolume = 0.75*Rgas*(T_K_f+T)/(P*P_Pa_f)*(1/V_c);
+
+  // define the local solver parameters
+  // for now, this is hard-coded, but we may want to let the user access the parameters at some point
+  integer const maxNumNewtonIter = 500;
+  integer const maxNumBacktrackIter = 20;
```

Comment:
This may be an extreme number of backtracks (0.5^20 ~ 1e-6 ~ zero update step).  It may be better to quit at 5-10 and just hope the next Newton step takes it to a better starting point.  Perhaps you can test on the challenging example to see if 20 are needed?

Reply:
You are right, thanks for catching this, that was too much backtracking. I reduced this number to 8 steps (max) and will give the user access to these values in a future PR.